File: src/facedetector.cpp
#include "facedetector.h"
#include <QDebug>
#include "logging.h"
#include <algorithm>
#include <cmath>

FaceDetector::FaceDetector(QObject *parent)
    : QObject(parent)
    , m_modelLoaded(false)
    , m_inputSize(640, 640)  // Fixed YuNet input size (good balance for mobile)
{
}

FaceDetector::~FaceDetector()
{
}

bool FaceDetector::loadModel(const QString &modelPath)
{
    try {
        qCDebug(lcNami) << "Loading YuNet face detection model from:" << modelPath;

        // Create FaceDetectorYN with OpenCV API
        // Parameters: model_path, config_path, input_size, score_threshold, nms_threshold, top_k, backend_id, target_id
        m_detector = cv::FaceDetectorYN::create(
            modelPath.toStdString(),  // model path
            "",                        // config path (empty for ONNX)
            cv::Size(m_inputSize.width(), m_inputSize.height()),  // input size
            0.6f,                      // score threshold (will be overridden in detect())
            0.3f,                      // NMS threshold
            5000,                      // top_k
            cv::dnn::DNN_BACKEND_OPENCV,   // backend
            cv::dnn::DNN_TARGET_CPU        // target
        );

        if (m_detector.empty()) {
            emit error("Failed to create YuNet detector");
            return false;
        }

        m_modelLoaded = true;
        qCDebug(lcNami) << "YuNet model loaded successfully";
        qCDebug(lcNami) << "Fixed input size:" << m_inputSize.width() << "x" << m_inputSize.height()
                 << "(all images will be resized to this)";

        return true;
    }
    catch (const cv::Exception &e) {
        QString errorMsg = QString("OpenCV exception loading model: %1").arg(e.what());
        qWarning() << errorMsg;
        emit error(errorMsg);
        return false;
    }
}

QVector<FaceDetection> FaceDetector::detect(const QImage &image, float confidenceThreshold)
{
    qCDebug(lcNami) << "QImage detection requested - size:" << image.width() << "x" << image.height();
    cv::Mat mat = qImageToCvMat(image);
    qCDebug(lcNami) << "Converted to cv::Mat - size:" << mat.cols << "x" << mat.rows;
    return detect(mat, confidenceThreshold);
}

QVector<FaceDetection> FaceDetector::detect(const cv::Mat &image, float confidenceThreshold)
{
    if (!m_modelLoaded) {
        emit error("Model not loaded");
        return QVector<FaceDetection>();
    }

    if (image.empty()) {
        emit error("Empty input image");
        return QVector<FaceDetection>();
    }

    qCDebug(lcNami) << "=== Face Detection Start ===";
    qCDebug(lcNami) << "Input image size:" << image.cols << "x" << image.rows << "channels:" << image.channels();
    qCDebug(lcNami) << "Confidence threshold:" << confidenceThreshold;

    try {
        // Downscale with a uniform factor so faces are not distorted; YuNet
        // supports arbitrary input sizes via setInputSize
        const int maxSide = qMax(m_inputSize.width(), m_inputSize.height());
        float scale = 1.0f;
        int longest = std::max(image.cols, image.rows);
        if (longest > maxSide) {
            scale = static_cast<float>(maxSide) / longest;
        }

        int newW = std::max(1, static_cast<int>(std::round(image.cols * scale)));
        int newH = std::max(1, static_cast<int>(std::round(image.rows * scale)));

        cv::Mat resizedImage;
        if (scale < 1.0f) {
            cv::resize(image, resizedImage, cv::Size(newW, newH), 0, 0, cv::INTER_AREA);
        } else {
            resizedImage = image;
        }

        m_detector->setInputSize(cv::Size(newW, newH));

        // Single uniform factor to map detections back to original coordinates
        float scaleX = static_cast<float>(image.cols) / newW;
        float scaleY = static_cast<float>(image.rows) / newH;

        qCDebug(lcNami) << "Resized to" << newW << "x" << newH << "scale:" << scaleX;

        // Set score threshold
        m_detector->setScoreThreshold(confidenceThreshold);

        // Detect faces on resized image
        cv::Mat faces;
        qCDebug(lcNami) << "Running YuNet detector...";
        m_detector->detect(resizedImage, faces);

        qCDebug(lcNami) << "Detection complete - faces matrix: rows=" << faces.rows << "cols=" << faces.cols << "type=" << faces.type();

        // Convert results
        QVector<FaceDetection> detections;

        if (faces.rows > 0) {
            qCDebug(lcNami) << "Found" << faces.rows << "faces";

            for (int i = 0; i < faces.rows; i++) {
                FaceDetection detection;

                // YuNet output format per row: [x, y, w, h, x_re, y_re, x_le, y_le, x_nt, y_nt, x_rcm, y_rcm, x_lcm, y_lcm, score]
                // Coordinates are in pixels

                float x = faces.at<float>(i, 0);
                float y = faces.at<float>(i, 1);
                float w = faces.at<float>(i, 2);
                float h = faces.at<float>(i, 3);
                float score = faces.at<float>(i, 14);

                qCDebug(lcNami) << "  Face" << i << "- bbox (pixels in detector input):" << x << y << w << h << "score:" << score;

                // Scale coordinates back to original image size
                float origX = x * scaleX;
                float origY = y * scaleY;
                float origW = w * scaleX;
                float origH = h * scaleY;

                qCDebug(lcNami) << "  Face" << i << "- bbox (pixels in original image):" << origX << origY << origW << origH;

                // Normalize to [0-1] based on original image size
                detection.bbox = QRectF(
                    origX / image.cols,
                    origY / image.rows,
                    origW / image.cols,
                    origH / image.rows
                );
                detection.confidence = score;

                // Extract 5 landmarks and normalize (also scale back to original)
                for (int j = 0; j < 5; j++) {
                    float lx = (faces.at<float>(i, 4 + j*2) * scaleX) / image.cols;
                    float ly = (faces.at<float>(i, 5 + j*2) * scaleY) / image.rows;
                    detection.landmarks.append(QPointF(lx, ly));
                }

                detections.append(detection);
            }
        } else {
            qCDebug(lcNami) << "No faces detected";
        }

        qCDebug(lcNami) << "=== Detection Complete: Found" << detections.size() << "faces ===";

        return detections;
    }
    catch (const cv::Exception &e) {
        QString errorMsg = QString("OpenCV exception during detection: %1").arg(e.what());
        qWarning() << errorMsg;
        emit error(errorMsg);
        return QVector<FaceDetection>();
    }
}

cv::Mat FaceDetector::qImageToCvMat(const QImage &image)
{
    // OpenCV convention is BGR; both YuNet and the recognition preprocessing
    // expect it, so everything downstream works on BGR mats
    QImage rgb = image.convertToFormat(QImage::Format_RGB888);
    cv::Mat rgbMat(rgb.height(), rgb.width(), CV_8UC3,
                   const_cast<uchar*>(rgb.bits()), rgb.bytesPerLine());

    cv::Mat bgr;
    cv::cvtColor(rgbMat, bgr, cv::COLOR_RGB2BGR);
    return bgr;
}

File: python/face_detector.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Face detection using YuNet model
Optimized for on-device processing on Sailfish OS
"""

import cv2
import numpy as np
import os
from pathlib import Path


class FaceDetector:
    """YuNet-based face detector for real-time face detection"""

    def __init__(self, model_path=None, conf_threshold=0.6, nms_threshold=0.3):
        """
        Initialize YuNet face detector

        Args:
            model_path: Path to YuNet ONNX model
            conf_threshold: Confidence threshold for detection (0.0-1.0)
            nms_threshold: NMS threshold for overlapping boxes
        """
        self.conf_threshold = conf_threshold
        self.nms_threshold = nms_threshold

        # Default model path
        if model_path is None:
            model_path = Path(__file__).parent / "models" / "face_detection_yunet_2023mar.onnx"

        if not os.path.exists(model_path):
            raise FileNotFoundError(f"YuNet model not found at {model_path}")

        # Initialize YuNet detector
        self.detector = cv2.FaceDetectorYN.create(
            str(model_path),
            "",
            (320, 320),  # Default input size
            conf_threshold,
            nms_threshold
        )

        self.input_size = (320, 320)

    def set_input_size(self, width, height):
        """Set input image size for detection"""
        self.input_size = (width, height)
        self.detector.setInputSize(self.input_size)

    def detect(self, image):
        """
        Detect faces in image

        Args:
            image: numpy array (BGR format)

        Returns:
            List of face detections, each containing:
            - bbox: [x, y, w, h]
            - confidence: float
            - landmarks: 5 facial landmarks (2 eyes, nose, 2 mouth corners)
        """
        if image is None or image.size == 0:
            return []

        height, width = image.shape[:2]

        # Adjust detector input size to match image
        if (width, height) != self.input_size:
            self.set_input_size(width, height)

        # Detect faces
        _, faces = self.detector.detect(image)

        if faces is None:
            return []

        # Parse detections
        detections = []
        for face in faces:
            # YuNet output format: [x, y, w, h, x_re, y_re, x_le, y_le, x_nt, y_nt, x_rcm, y_rcm, x_lcm, y_lcm, conf]
            # First 4: bounding box
            # Next 10: 5 landmarks (right eye, left eye, nose tip, right corner mouth, left corner mouth)
            # Last: confidence score

            bbox = face[:4].astype(int)
            landmarks = face[4:14].reshape(5, 2).astype(int)
            confidence = float(face[14])

            detection = {
                'bbox': bbox.tolist(),  # [x, y, w, h]
                'confidence': confidence,
                'landmarks': landmarks.tolist(),  # [[x, y], ...]
                'x': int(bbox[0]),
                'y': int(bbox[1]),
                'width': int(bbox[2]),
                'height': int(bbox[3])
            }

            detections.append(detection)

        return detections

    def detect_from_file(self, image_path):
        """
        Detect faces from image file

        Args:
            image_path: Path to image file

        Returns:
            List of face detections
        """
        image = cv2.imread(str(image_path))
        if image is None:
            raise ValueError(f"Could not read image from {image_path}")

        return self.detect(image)

    def visualize_detections(self, image, detections):
        """
        Draw bounding boxes and landmarks on image

        Args:
            image: numpy array (BGR format)
            detections: List of face detections

        Returns:
            Image with visualizations drawn
        """
        vis_image = image.copy()

        for det in detections:
            # Draw bounding box
            x, y, w, h = det['bbox']
            cv2.rectangle(vis_image, (x, y), (x + w, y + h), (0, 255, 0), 2)

            # Draw confidence
            conf_text = f"{det['confidence']:.2f}"
            cv2.putText(vis_image, conf_text, (x, y - 10),
                       cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 2)

            # Draw landmarks
            for landmark in det['landmarks']:
                cv2.circle(vis_image, tuple(landmark), 2, (0, 0, 255), -1)

        return vis_image

    def crop_face(self, image, detection, padding=0.2):
        """
        Crop face region from image with padding

        Args:
            image: numpy array (BGR format)
            detection: Single face detection dict
            padding: Padding ratio around face (0.0-1.0)

        Returns:
            Cropped face image
        """
        x, y, w, h = detection['bbox']

        # Add padding
        pad_w = int(w * padding)
        pad_h = int(h * padding)

        x1 = max(0, x - pad_w)
        y1 = max(0, y - pad_h)
        x2 = min(image.shape[1], x + w + pad_w)
        y2 = min(image.shape[0], y + h + pad_h)

        face_crop = image[y1:y2, x1:x2]

        return face_crop

    def align_face(self, image, detection, output_size=(112, 112)):
        """
        Align face using landmarks (for better recognition)

        Args:
            image: numpy array (BGR format)
            detection: Single face detection dict
            output_size: Size of output aligned face

        Returns:
            Aligned face image
        """
        landmarks = np.array(detection['landmarks'], dtype=np.float32)

        # Use eye landmarks for alignment
        left_eye = landmarks[0]  # Right eye in image (left eye of person)
        right_eye = landmarks[1]  # Left eye in image (right eye of person)

        # Calculate angle
        dy = right_eye[1] - left_eye[1]
        dx = right_eye[0] - left_eye[0]
        angle = np.degrees(np.arctan2(dy, dx))

        # Get center point between eyes
        eye_center = ((left_eye[0] + right_eye[0]) / 2,
                     (left_eye[1] + right_eye[1]) / 2)

        # Get rotation matrix
        M = cv2.getRotationMatrix2D(eye_center, angle, 1.0)

        # Rotate image
        aligned = cv2.warpAffine(image, M, (image.shape[1], image.shape[0]))

        # Crop and resize to output size
        face_crop = self.crop_face(aligned, detection, padding=0.3)
        face_resized = cv2.resize(face_crop, output_size)

        return face_resized


if __name__ == "__main__":
    # Test detector
    print("YuNet Face Detector Test")
    print("=" * 50)

    # Initialize detector
    detector = FaceDetector()
    print(f"Detector initialized")
    print(f"Confidence threshold: {detector.conf_threshold}")
    print(f"NMS threshold: {detector.nms_threshold}")
